### PR TITLE
UCT/UD: Fix issue N544 (CREQ assert)

### LINKED ISSUE #544: uct_ud_ep_rx_creq error at np 1280
commit 5e4eb9eeb667b79b5f2ab52c7658fae5e2a43cfc

Hey guys, at np 1280 I'm seeing the following error, is this a known issue?
I was testing IMB-MPI1 benchmarks using pml_ucx from ompi-trunk.

```
$mpirun -np 1280 --map-by ppr:20:node -mca pml ucx -x UCX_TLS=rc_x -x UCX_DEVICES=mlx5_0:1 ./IMB-MPI1 -npmin 1280

[clx-orion-038:8784 :0]       ud_ep.c:315  Assertion `ctl->type == UCT_UD_PACKET_CREQ' failed

/labhome/tomislavj/scrap/ucx-trunk/src/uct/ib/ud/base/ud_ep.c: [ uct_ud_ep_rx_creq() ]
  ...
  312     uct_ud_ctl_hdr_t *ctl = (uct_ud_ctl_hdr_t *)(neth + 1);
  313
  314     ucs_assert_always(ctl->type == UCT_UD_PACKET_CREQ);
==>   315
  316     ep = uct_ud_iface_cep_lookup(iface, &ctl->conn_req.ib_addr, ctl->conn_req.conn_id);
  317     if (!ep) {
  318         ep = uct_ud_ep_create_passive(iface, ctl);

==== backtrace ====
 0 0x000000000001d961 uct_ud_ep_rx_creq()  /labhome/tomislavj/scrap/ucx-trunk/src/uct/ib/ud/base/ud_ep.c:315
 1 0x000000000001d961 uct_ud_ep_rx_creq()  /labhome/tomislavj/scrap/ucx-trunk/src/uct/ib/ud/base/ud_ep.c:319
 2 0x000000000001d961 uct_ud_ep_process_rx()  /labhome/tomislavj/scrap/ucx-trunk/src/uct/ib/ud/base/ud_ep.c:407
 3 0x000000000001ee84 uct_ud_verbs_iface_poll_rx()  /labhome/tomislavj/scrap/ucx-trunk/src/uct/ib/ud/verbs/ud_verbs.c:286
 4 0x000000000001ee84 uct_ud_verbs_iface_progress()  /labhome/tomislavj/scrap/ucx-trunk/src/uct/ib/ud/verbs/ud_verbs.c:315
 5 0x000000000000ef1a ucs_notifier_chain_call()  /labhome/tomislavj/scrap/ucx-trunk/src/ucs/datastruct/notifier.h:52
 6 0x000000000000ef1a uct_worker_progress()  /labhome/tomislavj/scrap/ucx-trunk/src/uct/base/uct_pd.c:210
 7 0x000000000000734d ucp_worker_progress()  /labhome/tomislavj/scrap/ucx-trunk/src/ucp/core/ucp_worker.c:230
 8 0x000000000000734d ucs_async_check_miss()  /labhome/tomislavj/scrap/ucx-trunk/src/ucs/async/async.h:135
 9 0x000000000000734d ucp_worker_progress()  /labhome/tomislavj/scrap/ucx-trunk/src/ucp/core/ucp_worker.c:231
10 0x0000000000002ddb mca_pml_ucx_progress()  /labhome/tomislavj/scrap/ompi-trunk/ompi/mca/pml/ucx/pml_ucx.c:323
11 0x000000000002f5ba opal_progress()  /labhome/tomislavj/scrap/ompi-trunk/opal/runtime/opal_progress.c:189
12 0x0000000000044d1d opal_condition_wait()  /labhome/tomislavj/scrap/ompi-trunk/ompi/../opal/threads/condition.h:76
13 0x0000000000044d1d ompi_request_default_wait_all()  /labhome/tomislavj/scrap/ompi-trunk/ompi/request/req_wait.c:287
14 0x0000000000072178 ompi_coll_base_sendrecv_nonzero_actual()  /labhome/tomislavj/scrap/ompi-trunk/ompi/mca/coll/base/coll_base_util.c:66
15 0x0000000000070a52 ompi_coll_base_sendrecv()  /labhome/tomislavj/scrap/ompi-trunk/ompi/mca/coll/base/coll_base_util.h:67
16 0x000000000002c6ad ompi_comm_split()  /labhome/tomislavj/scrap/ompi-trunk/ompi/communicator/comm.c:457
17 0x0000000000057929 PMPI_Comm_split()  /labhome/tomislavj/scrap/ompi-trunk/ompi/mpi/c/profile/pcomm_split.c:69
18 0x0000000000402ad9 IMB_set_communicator()  ??:0
19 0x0000000000402b27 IMB_init_communicator()  ??:0
20 0x0000000000401f17 main()  ??:0
21 0x0000003c8fe1ed1d __libc_start_main()  ??:0
22 0x0000000000401ce9 _start()  ??:0
===================
```

### PR DESCRIPTION
**Pull request by brminich**
Fixed:
 - If CREQ and CREP to peer are lost, peer may stay unconnected
 - type == UCT_UD_PACKET_CREQ assertion

fixes #544

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2337/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1667/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@alex-mikheev, could you please look thru?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1668/ for details (Mellanox internal link).

File: src/uct/ib/ud/base/ud_ep.c [hunk 1]
```diff
@@ -578,6 +579,9 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
             goto out;
         }
         if (neth->packet_type & UCT_UD_PACKET_FLAG_CTL) {
+            /* No need to track duplications, CREP always goes
+             * with ACK_REQ flag */
+            ep->rx.ooo_pkts.head_sn = neth->psn;
             uct_ud_ep_rx_ctl(iface, ep, (uct_ud_ctl_hdr_t *)(neth + 1));
```
Comment (alex-mikheev) [inline src/uct/ib/ud/base/ud_ep.c]: is it possible to receive creq here ?

Comment (alex-mikheev) [inline src/uct/ib/ud/base/ud_ep.c]: it think it is better to add this line in uct_ud_ep_rx_ctl()

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1672/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2341/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1674/ for details (Mellanox internal link).

**Comment by alex-mikheev** [comment]
:+1:

**Pull request merged**